services/supervisor/src/lib.rs

```diff
@@ -213,53 +204,66 @@ fn assign_instance_id(context: &CallContext<'_>) -> InstanceId {
 
 /// Supervisor service implementation.
 #[derive(Debug, Default, Clone, ServiceFactory, ServiceDispatcher)]
-#[service_dispatcher(implements("transactions::SupervisorInterface"))]
-#[service_factory(
-    proto_sources = "proto",
-    artifact_name = "exonum-supervisor",
-    service_constructor = "Self::construct"
-)]
-pub struct Supervisor<Mode>
-where
-    Mode: mode::SupervisorMode,
-{
-    phantom: std::marker::PhantomData<Mode>,
-}
-
-impl<Mode> Supervisor<Mode>
-where
-    Mode: mode::SupervisorMode,
-{
+#[service_dispatcher(implements(
+    "transactions::SupervisorInterface",
+    "Configure<Params = SupervisorConfig>"
+))]
+#[service_factory(proto_sources = "proto", artifact_name = "exonum-supervisor")]
+pub struct Supervisor;
+
+impl Supervisor {
     /// Name of the supervisor service.
     pub const NAME: &'static str = "supervisor";
 
-    /// Creates a new `Supervisor` service object.
-    pub fn new() -> Supervisor<Mode> {
-        Supervisor {
-            phantom: std::marker::PhantomData::<Mode>::default(),
+    /// Creates a configuration for a simple `Supervisor`.
+    pub fn simple_config() -> SupervisorConfig {
+        SupervisorConfig { mode: Mode::Simple }
+    }
+
+    /// Creates a configuration for a decentralized `Supervisor`.
+    pub fn decentralized_config() -> SupervisorConfig {
+        SupervisorConfig {
+            mode: Mode::Decentralized,
         }
     }
 
-    /// Factory constructor of the `Supervisor` object that takes `&self` as an argument.
-    /// The constructor is required for the `ServiceFactory` trait implementation.
-    pub fn construct(&self) -> Box<Self> {
-        Box::new(Self::new())
+    /// Creates an `InstanceCollection` with builtin `Supervisor` instance given the
+    /// configuration.
+    pub fn builtin_instance(config: SupervisorConfig) -> InstanceInitParams {
+        Supervisor
+            .artifact_id()
+            .into_default_instance(SUPERVISOR_INSTANCE_ID, Self::NAME)
+            .with_constructor(config)
     }
 }
 
-impl<Mode> Service for Supervisor<Mode>
-where
-    Mode: mode::SupervisorMode,
-{
+impl Service for Supervisor {
+    fn initialize(&self, context: CallContext<'_>, params: Vec<u8>) -> Result<(), ExecutionError> {
+        use std::borrow::Cow;
+
+        // Load configuration from bytes and store it.
+        let config = SupervisorConfig::from_bytes(Cow::from(&params)).unwrap_or_else(|err| {
+            // Incorrect config is a critical error for both the service and the
```

Comment:
I don't think we should think about future-proofing (namely, dynamically added / multiple supervisors in this case), but I want to draw your attention that an error in a builtin service constructor [bubbles up](https://github.com/exonum/exonum/blob/88a782fef9d32676e483033d19acc8f37ecf1c28/exonum/src/blockchain/mod.rs#L243). Thus, panicking seems sort of redundant.